src/account/PluginManager2.sol

```diff
@@ -6,63 +6,109 @@ import {EnumerableSet} from "@openzeppelin/contracts/utils/structs/EnumerableSet
 import {IPlugin} from "../interfaces/IPlugin.sol";
 import {FunctionReference} from "../interfaces/IPluginManager.sol";
 import {FunctionReferenceLib} from "../helpers/FunctionReferenceLib.sol";
-import {AccountStorage, getAccountStorage, toSetValue} from "./AccountStorage.sol";
+import {AccountStorage, getAccountStorage, toSetValue, toFunctionReference} from "./AccountStorage.sol";
 
 // Temporary additional functions for a user-controlled install flow for validation functions.
 abstract contract PluginManager2 {
     using EnumerableSet for EnumerableSet.Bytes32Set;
 
-    error DefaultValidationAlreadySet(address plugin, uint8 functionId);
-    error ValidationAlreadySet(bytes4 selector, address plugin, uint8 functionId);
-    error ValidationNotSet(bytes4 selector, address plugin, uint8 functionId);
+    error DefaultValidationAlreadySet(FunctionReference validationFunction);
+    error ValidationAlreadySet(bytes4 selector, FunctionReference validationFunction);
+    error PreValidationAlreadySet(FunctionReference validationFunction, FunctionReference preValidationFunction);
+    error ValidationNotSet(bytes4 selector, FunctionReference validationFunction);
 
     function _installValidation(
-        address plugin,
-        uint8 functionId,
+        FunctionReference validationFunction,
         bool isDefault,
         bytes4[] memory selectors,
-        bytes calldata installData
-    ) internal {
-        FunctionReference validationFunction = FunctionReferenceLib.pack(plugin, functionId);
-
+        bytes calldata installData,
+        bytes memory preValidationHooks
+    )
+        // TODO: flag for signature validation
+        internal
+    {
         AccountStorage storage _storage = getAccountStorage();
 
+        if (preValidationHooks.length > 0) {
+            (FunctionReference[] memory preValidationFunctions, bytes[] memory initDatas) =
+                abi.decode(preValidationHooks, (FunctionReference[], bytes[]));
+
+            for (uint256 i = 0; i < preValidationFunctions.length; ++i) {
+                FunctionReference preValidationFunction = preValidationFunctions[i];
+
+                if (
+                    !_storage.validationData[validationFunction].preValidationHooks.add(
+                        toSetValue(preValidationFunction)
+                    )
+                ) {
+                    revert PreValidationAlreadySet(validationFunction, preValidationFunction);
+                }
+
+                if (initDatas[i].length > 0) {
+                    (address preValidationPlugin,) = FunctionReferenceLib.unpack(preValidationFunction);
+                    IPlugin(preValidationPlugin).onInstall(initDatas[i]);
```

Comment:
So dependencies no longer need to be installed ahead of time? Do we expect the plugin to be able to handle multiple `onInstall`s with different data?

Side note: It's interesting how we're going back towards the `updatePlugins` pattern that we had before. 😅

Reply:
This workflow kinda doesn't touch dependencies at all, since it's not reading from the manifest 😅 .

As for expectations around `onInstall`, this function currently allows the caller to decide whether or not to send it multiple times - if no data is provided, the `onInstall` call is skipped. Not sure if that should be the final design or not.

Reply:
Re: dependencies - I meant in the sense that we had previously operated under the assumption that plugins can only use what is available from plugins installed already. This sort of breaks that and allows us to pull in arbitrary hooks from non-installed plugins.

Reply:
Not saying the previous approach is right though. I've also been thinking about what it would look like to get rid of the idea of "installing" plugins, and just "using" them instead.

Reply:
Ah, I see what you mean. In a sense, this function is doing the same for the validation function too - the process of "installing" just becomes setting the storage in the account to allow using it, and optionally sending the `onInstall` callback with some data.

It does create a discrepancy with how the account currently treats a few things related to plugin install state:
- The implementing plugins for both the validation function and pre-validation hooks won't appear in the list of installed plugins from the loupe function `getInstalledPlugins()`.
- The validation functions can't be used as dependencies, because the implementing plugin isn't held on that list.
- The manifest hash is not stored (nor is the manifest even checked)

These all seem like things we should follow up on. It may not be useful for us to even have a `getInstalledPlugins` function anymore, if they can be installed multiple times in different circumstances. And the extent of manifest contents and checking needs an overhaul in user-supplied install configs.

Reply:
😅 

I think all worth following up on for sure. We're no longer getting value out of the integrity of the manifest. Happy to have us revisit in a future PR. We're deep into this approach already.